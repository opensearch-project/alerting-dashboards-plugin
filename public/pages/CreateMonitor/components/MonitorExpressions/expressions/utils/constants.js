/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const POPOVER_STYLE = { zIndex: '200' };
export const EXPRESSION_STYLE = { padding: '20px', whiteSpace: 'nowrap' };
export const Expressions = {
  THRESHOLD: 'THRESHOLD',
  WHEN: 'WHEN',
  OF_FIELD: 'OF_FIELD',
  OVER: 'OVER',
  FOR_THE_LAST: 'FOR_THE_LAST',
  WHERE: 'WHERE',
  METRICS: 'METRICS',
  GROUP_BY: 'GROUP_BY',
};
export const NUMBER_TYPES = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
];
export const UNITS_OF_TIME = [
  { value: 'm', text: 'minute(s)' },
  { value: 'h', text: 'hour(s)' },
  { value: 'd', text: 'day(s)' },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: 'Select value', value: '' },
  { text: 'True', value: true },
  { text: 'False', value: false },
];

export const OPERATORS_MAP = {
  IS: {
    text: 'is',
    value: 'is',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT: {
    text: 'is not',
    value: 'is_not',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NULL: {
    text: 'is null',
    value: 'is_null',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT_NULL: {
    text: 'is not null',
    value: 'is_not_null',
    dataTypes: ['number', 'text', 'keyword'],
  },
  IS_GREATER: {
    text: 'is greater than',
    value: 'is_greater',
    dataTypes: ['number'],
  },
  IS_GREATER_EQUAL: {
    text: 'is greater than equal',
    value: 'is_greater_equal',
    dataTypes: ['number'],
  },
  IS_LESS: {
    text: 'is less than',
    value: 'is_less',
    dataTypes: ['number'],
  },
  IS_LESS_EQUAL: {
    text: 'is less than equal',
    value: 'is_less_equal',
    dataTypes: ['number'],
  },
  STARTS_WITH: {
    text: 'starts with',
    value: 'starts_with',
    dataTypes: ['text', 'keyword'],
  },
  ENDS_WITH: {
    text: 'ends with',
    value: 'ends_with',
    dataTypes: ['text', 'keyword'],
  },
  CONTAINS: {
    text: 'contains',
    value: 'contains',
    dataTypes: ['text', 'keyword'],
  },
  DOES_NOT_CONTAINS: {
    text: 'does not contain',
    value: 'does_not_contains',
    dataTypes: ['text'],
  },
  IN_RANGE: {
    text: 'is in range',
    value: 'in_range',
    dataTypes: ['number'],
  },
  NOT_IN_RANGE: {
    text: 'is not in range',
    value: 'not_in_range',
    dataTypes: ['number'],
  },
};

export const OVER_TYPES = [{ value: 'all documents', text: 'all documents' }];

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const GROUP_BY_ERROR = 'Must specify at least 1 group by expression.';
export const QUERY_TYPE_GROUP_BY_ERROR = 'Can have a maximum of 1 group by selections.';

export const QUERY_TYPE_METRIC_ERROR = 'Can have a maximum of 1 metric selections.';

export const MAX_NUM_QUERY_LEVEL_GROUP_BYS = 1;
export const MAX_NUM_BUCKET_LEVEL_GROUP_BYS = 2;
export const MAX_NUM_WHERE_EXPRESSION = 1;
export const WHERE_FILTER_ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];
