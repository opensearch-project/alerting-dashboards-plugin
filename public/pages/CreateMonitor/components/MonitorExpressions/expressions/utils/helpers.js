/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export function selectOptionValueToText(optionValue, options) {
  return options.find((opt) => opt.value === optionValue).text;
}

export function getOfExpressionAllowedTypes(values) {
  const types = ['number'];
  if (['min', 'max'].includes(values.aggregationType)) types.push('date');
  return types;
}

export function getMetricExpressionAllowedTypes(values) {
  const types = ['number'];

  // Check if the aggregation types includes any of 'min', 'max', or 'count'.
  // If so, add 'date' to the allowed type for metric expression field.
  if (
    values.aggregations.some((e) => {
      return ['min', 'max', 'count'].includes(e.aggregationType);
    })
  )
    types.push('date');

  // Check if the aggregations types includes 'count'.
  // If so, add 'keyword' to the allowed type for metric expression field.
  if (
    values.aggregations.some((e) => {
      return ['count'].includes(e.aggregationType);
    })
  )
    types.push('keyword');
  return types;
}

export function getGroupByExpressionAllowedTypes() {
  return ['keyword'];
}

export const validateAggregationsDuplicates = (aggregations) => {
  let duplicates = 0;
  aggregations.forEach((e1, index) => {
    aggregations.slice(index + 1).forEach((e2) => {
      if (e1.aggregationType === e2.aggregationType && e1.fieldName === e2.fieldName) {
        duplicates++;
      }
    });
  });
  return duplicates > 0;
};
