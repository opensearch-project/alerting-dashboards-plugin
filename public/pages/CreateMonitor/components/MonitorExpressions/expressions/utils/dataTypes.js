/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { NUMBER_TYPES } from './constants';
import { DATA_TYPES } from '../../../../../../utils/constants';

export function getFieldsForType(dataTypes, type) {
  if (type === DATA_TYPES.NUMBER) {
    return NUMBER_TYPES.reduce(
      (options, type) => options.concat(getFieldsForType(dataTypes, type)),
      []
    );
  }
  if (dataTypes[type]) {
    return [...dataTypes[type].values()];
  }
  return [];
}

export const getIndexFields = (dataTypes, allowedTypes) =>
  allowedTypes.map((type) => ({
    label: type,
    options: getFieldsForType(dataTypes, type).map((field) => ({ label: field, type })),
  }));

export const getFilteredIndexFields = (dataTypes, allowedTypes, fieldsToInclude) =>
  allowedTypes.map((type) => ({
    label: type,
    options: getFieldsForType(dataTypes, type)
      .filter((field) => fieldsToInclude.includes(field))
      .map((field) => ({ label: field, type })),
  }));

/**
 * Searches the output from 'getIndexFields' for the provided field to determine the field's type.
 * @param queryField - a string representing the field being queried.
 * @param indexFields - the array output from 'getIndexFields'.
 * @return {*|string} - a string representing the data type of the queryField.
 */
export function getTypeForField(queryField = '', indexFields = []) {
  if (_.isEmpty(queryField) || _.isEmpty(indexFields)) return DATA_TYPES.TEXT;
  const match = indexFields.find(
    (entry) => entry.options?.find((dataField) => dataField.label === queryField) !== undefined
  );
  return match?.label || DATA_TYPES.TEXT;
}
