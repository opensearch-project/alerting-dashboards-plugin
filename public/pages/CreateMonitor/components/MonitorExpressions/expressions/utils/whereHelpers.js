/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { OPERATORS_MAP, Expressions } from './constants';
import { TRIGGER_COMPARISON_OPERATORS } from '../../../../../CreateTrigger/containers/DefineBucketLevelTrigger/DefineBucketLevelTrigger';
import { DATA_TYPES } from '../../../../../../utils/constants';
import { FORMIK_INITIAL_WHERE_EXPRESSION_VALUES } from '../../../../containers/CreateMonitor/utils/constants';

export const DEFAULT_WHERE_EXPRESSION_TEXT = 'All fields are included';

export const getOperators = (
  fieldType = DATA_TYPES.TEXT,
  supportedOperators = Object.values(OPERATORS_MAP)
) =>
  supportedOperators.reduce(
    (acc, currentOperator) =>
      currentOperator.dataTypes.includes(fieldType)
        ? [...acc, { text: currentOperator.text, value: currentOperator.value }]
        : acc,
    []
  );

export const isRangeOperator = (selectedOperator) =>
  [OPERATORS_MAP.IN_RANGE.value, OPERATORS_MAP.NOT_IN_RANGE.value].includes(selectedOperator);
export const isNullOperator = (selectedOperator) =>
  [OPERATORS_MAP.IS_NULL.value, OPERATORS_MAP.IS_NOT_NULL.value].includes(selectedOperator);

export const displayText = (whereValues) => {
  const comparisonOperators = _.concat(Object.values(OPERATORS_MAP), TRIGGER_COMPARISON_OPERATORS);

  const whereFieldName = _.get(whereValues, 'fieldName[0].label', undefined);
  if (!whereFieldName) return DEFAULT_WHERE_EXPRESSION_TEXT;

  const selectedOperator = _.get(whereValues, 'operator', OPERATORS_MAP.IS);
  const operatorObj =
    comparisonOperators.find((operator) => operator.value === selectedOperator) || {};
  const initialText = `${whereFieldName} ${operatorObj.text || ''}`;

  if (isRangeOperator(selectedOperator)) {
    const startRange = _.get(whereValues, 'fieldRangeStart');
    const endRange = _.get(whereValues, 'fieldRangeEnd');
    return `${initialText} from ${startRange} to ${endRange}`;
  } else if (isNullOperator(selectedOperator)) {
    return `${initialText}`;
  } else {
    const value = _.get(whereValues, 'fieldValue', '');
    return `${initialText} ${value}`;
  }
};

export const validateRange = (value = '', filter = FORMIK_INITIAL_WHERE_EXPRESSION_VALUES) => {
  if (_.isEmpty(value?.toString())) return 'Required.';
  if (filter.fieldRangeEnd < value || filter.fieldRangeStart === filter.fieldRangeEnd)
    return 'Start should be less than end range.';
  if (value < filter.fieldRangeStart) return 'End should be greater than start range.';
  return undefined;
};

export const validateWhereFilter = (filter = FORMIK_INITIAL_WHERE_EXPRESSION_VALUES) => {
  const fieldName = _.get(
    filter,
    'fieldName[0]',
    FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.fieldName[0]
  );
  const fieldOperator = _.get(filter, 'operator', FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.operator);
  let filterIsValid = !_.isEmpty(fieldName.label);
  switch (fieldOperator) {
    case OPERATORS_MAP.IS:
    case OPERATORS_MAP.IS_NOT:
    case OPERATORS_MAP.IS_GREATER:
    case OPERATORS_MAP.IS_GREATER_EQUAL:
    case OPERATORS_MAP.IS_LESS:
    case OPERATORS_MAP.IS_LESS_EQUAL:
    case OPERATORS_MAP.STARTS_WITH:
    case OPERATORS_MAP.ENDS_WITH:
    case OPERATORS_MAP.CONTAINS:
    case OPERATORS_MAP.NOT_CONTAINS:
      // These operators store the query value in the 'fieldValue'
      // attribute of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
      // Validate that value.
      filterIsValid = filterIsValid && !_.isEmpty(filter.fieldValue?.toString());
      break;
    case OPERATORS_MAP.IN_RANGE:
    case OPERATORS_MAP.NOT_IN_RANGE:
      // These operators store the query values in the 'fieldRangeStart' and 'fieldRangeEnd'
      // attributes of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
      // Both of those values need to be validated.
      filterIsValid =
        filterIsValid &&
        !validateRange(filter.fieldRangeStart, filter) &&
        !validateRange(filter.fieldRangeEnd, filter);
      break;
    case OPERATORS_MAP.IS_NULL:
    case OPERATORS_MAP.IS_NOT_NULL:
      // These operators don't store a query value in the FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
      // No further validation needed.
      break;
    default:
      console.log('Unknown query operator detected:', fieldOperator);
      filterIsValid = false;
  }
  return filterIsValid;
};

export const validateWhereFilters = (filters = []) => {
  return filters.filter((filter) => validateWhereFilter(filter)).length === filters.length;
};

export const getIsDataFilterActive = ({ formik: { values }, openedStates, fieldPath = '' }) => {
  const whereValues = _.get(values, `${fieldPath}where`);
  const whereFieldName = _.get(whereValues, 'fieldName[0].label', undefined);
  return openedStates[Expressions.WHERE] || whereFieldName;
};
