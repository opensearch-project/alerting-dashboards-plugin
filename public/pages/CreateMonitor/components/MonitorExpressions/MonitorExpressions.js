/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from 'react';
import { EuiSpacer } from '@elastic/eui';
import { connect } from 'formik';
import { ForExpression, WhereExpression } from './expressions';
import MetricExpression from './expressions/MetricExpression';
import { FieldArray } from 'formik';
import GroupByExpression from './expressions/GroupByExpression';
import MinimalAccordion from '../../../../components/FeatureAnywhereContextMenu/MinimalAccordion';
import {
  UNITS_OF_TIME,
  MAX_NUM_QUERY_LEVEL_GROUP_BYS,
  MAX_NUM_BUCKET_LEVEL_GROUP_BYS,
  MAX_NUM_WHERE_EXPRESSION,
} from './expressions/utils/constants';
import { getIsDataFilterActive } from './expressions/utils/whereHelpers';
import { MONITOR_TYPE } from '../../../../utils/constants';

export const DEFAULT_CLOSED_STATES = {
  WHERE: false,
  // not using
  METRICS: false,
  GROUP_BY: false,
  OVER: false,
  FOR_THE_LAST: false,
  THRESHOLD: false,
  WHEN: false,
  OF_FIELD: false,
};

class MonitorExpressions extends Component {
  constructor(props) {
    super(props);

    this.state = {
      openedStates: DEFAULT_CLOSED_STATES,
      madeChanges: false,
      accordionsOpen: {},
      SectionContainer: props.flyoutMode ? MinimalAccordion : ({ children }) => <>{children}</>,
    };
  }

  openExpression = (expression) => {
    this.setState({
      openedStates: {
        ...DEFAULT_CLOSED_STATES,
        [expression]: true,
      },
    });
  };

  closeExpression = (expression) => {
    const { madeChanges, openedStates } = this.state;
    if (madeChanges && openedStates[expression]) {
      // if made changes and close expression that was currently open => run query
      this.setState({ madeChanges: false });
    }
    this.setState({ openedStates: { ...openedStates, [expression]: false } });
  };

  onMadeChanges = () => {
    this.setState({ madeChanges: true });
  };

  getExpressionProps = () => ({
    openedStates: this.state.openedStates,
    closeExpression: this.closeExpression,
    openExpression: this.openExpression,
    onMadeChanges: this.onMadeChanges,
  });

  onAccordionToggle = (key) => {
    const accordionsOpen = { [key]: !this.state.accordionsOpen[key] };
    this.setState({ accordionsOpen });
  };

  render() {
    const { accordionsOpen, SectionContainer } = this.state;
    const { dataTypes, errors, flyoutMode, formik } = this.props;
    const unit = UNITS_OF_TIME.find(({ value }) => value === formik?.values?.bucketUnitOfTime);
    const expressionProps = this.getExpressionProps();
    const isDataFilterActive = getIsDataFilterActive({ formik, ...expressionProps });
    const isBucketLevelMonitor = formik.values.monitor_type === MONITOR_TYPE.BUCKET_LEVEL;
    const groupBysLimit = isBucketLevelMonitor
      ? MAX_NUM_BUCKET_LEVEL_GROUP_BYS
      : MAX_NUM_QUERY_LEVEL_GROUP_BYS;

    return (
      <div>
        <SectionContainer
          {...{
            title: 'Metrics',
            subTitle: 'COUNT OF documents. You can add up to 1 metric.',
            id: 'metric-expression__metrics',
            isOpen: accordionsOpen.metrics,
            onToggle: () => this.onAccordionToggle('metrics'),
          }}
        >
          <FieldArray name="aggregations" validateOnChange={false}>
            {(arrayHelpers) => (
              <MetricExpression
                errors={errors}
                arrayHelpers={arrayHelpers}
                dataTypes={dataTypes}
                flyoutMode={flyoutMode}
              />
            )}
          </FieldArray>
        </SectionContainer>
        <EuiSpacer size={flyoutMode ? 'xs' : 'm'} />
        <SectionContainer
          {...{
            title: 'Time range',
            subTitle: `Last ${formik.values.bucketValue} ${unit?.text}`,
            isUsingDivider: true,
            id: 'metric-expression__time-range',
            isOpen: accordionsOpen.timeRange,
            onToggle: () => this.onAccordionToggle('timeRange'),
          }}
        >
          <ForExpression />
        </SectionContainer>
        <EuiSpacer size={flyoutMode ? 'xs' : 'l'} />
        <SectionContainer
          {...{
            title: 'Data filter',
            subTitle: isDataFilterActive
              ? '1 filter defined'
              : `No filter defined. You can add up to ${MAX_NUM_WHERE_EXPRESSION} filter.`,
            isUsingDivider: true,
            id: 'metric-expression__data-filter',
            isOpen: accordionsOpen.dataFilter,
            onToggle: () => this.onAccordionToggle('dataFilter'),
          }}
        >
          <WhereExpression {...expressionProps} dataTypes={dataTypes} flyoutMode={flyoutMode} />
        </SectionContainer>
        <EuiSpacer size={flyoutMode ? 'xs' : 'm'} />
        <SectionContainer
          {...{
            title: 'Group by',
            subTitle: formik.values.groupBy.length
              ? `${formik.values.groupBy.length} groupbys defined`
              : `No group bys defined. You can add up to ${groupBysLimit} group bys.`,
            isUsingDivider: true,
            id: 'metric-expression__group-by',
            isOpen: accordionsOpen.groupBy,
            onToggle: () => this.onAccordionToggle('groupBy'),
          }}
        >
          <FieldArray name="groupBy" validateOnChange={false}>
            {(arrayHelpers) => (
              <GroupByExpression
                errors={errors}
                arrayHelpers={arrayHelpers}
                dataTypes={dataTypes}
                flyoutMode={flyoutMode}
              />
            )}
          </FieldArray>
        </SectionContainer>
        <EuiSpacer size="xs" />
      </div>
    );
  }
}

export default connect(MonitorExpressions);
