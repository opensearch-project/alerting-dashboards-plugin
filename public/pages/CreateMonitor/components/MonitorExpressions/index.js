/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import MonitorExpressions from './MonitorExpressions';

export default MonitorExpressions;
