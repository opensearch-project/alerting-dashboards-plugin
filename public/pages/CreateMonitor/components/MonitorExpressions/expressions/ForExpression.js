/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component } from 'react';
import { EuiText, EuiFlexGroup, EuiFlexItem, EuiSpacer } from '@elastic/eui';
import { UNITS_OF_TIME } from './utils/constants';
import { FormikFieldNumber, FormikSelect } from '../../../../../components/FormControls';
import { hasError, isInvalid, validatePositiveInteger } from '../../../../../utils/validate';

class ForExpression extends Component {
  render() {
    return (
      <div>
        <EuiText size="xs">
          <strong>Time range for the last</strong>
        </EuiText>
        <EuiSpacer size={'s'} />
        <EuiFlexGroup style={{ maxWidth: 600 }}>
          <EuiFlexItem grow={false} style={{ width: 150 }}>
            <FormikFieldNumber
              name="bucketValue"
              formRow
              fieldProps={{ validate: validatePositiveInteger }}
              inputProps={{ min: 1, onChange: this.onChangeWrapper }}
              rowProps={{
                isInvalid,
                error: hasError,
              }}
            />
          </EuiFlexItem>
          <EuiFlexItem grow={false} style={{ width: 150 }}>
            <FormikSelect
              name="bucketUnitOfTime"
              inputProps={{
                options: UNITS_OF_TIME,
              }}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
      </div>
    );
  }
}

export default ForExpression;
