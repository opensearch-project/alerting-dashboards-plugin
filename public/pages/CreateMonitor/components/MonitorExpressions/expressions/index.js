/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ForExpression from './ForExpression';
import GroupByExpression from './GroupByExpression';
import GroupByItem from './GroupByItem';
import GroupByPopover from './GroupByPopover';
import MetricExpression from './MetricExpression';
import MetricItem from './MetricItem';
import MetricPopover from './MetricPopover';
import OfExpression from './OfExpression';
import OverExpression from './OverExpression';
import WhenExpression from './WhenExpression';
import WhereExpression from './WhereExpression';
import WhereExpressionFlyout from './WhereExpressionFlyout';

export {
  ForExpression,
  GroupByExpression,
  GroupByItem,
  GroupByPopover,
  MetricExpression,
  MetricItem,
  MetricPopover,
  OfExpression,
  OverExpression,
  WhenExpression,
  WhereExpression,
  WhereExpressionFlyout,
};
