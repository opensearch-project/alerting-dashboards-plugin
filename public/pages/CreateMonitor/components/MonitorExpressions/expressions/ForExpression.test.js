/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Formik } from 'formik';
import { render } from 'enzyme';

import { FORMIK_INITIAL_VALUES } from '../../../containers/CreateMonitor/utils/constants';
import ForExpression from './ForExpression';
import { DEFAULT_CLOSED_STATES } from '../MonitorExpressions';

describe('ForExpression', () => {
  test('renders', () => {
    const component = (
      <Formik initialValues={FORMIK_INITIAL_VALUES}>
        {(props) => (
          <ForExpression
            formik={props}
            openedStates={DEFAULT_CLOSED_STATES}
            openExpression={() => {}}
            closeExpression={() => {}}
          />
        )}
      </Formik>
    );

    expect(render(component)).toMatchSnapshot();
  });
});
