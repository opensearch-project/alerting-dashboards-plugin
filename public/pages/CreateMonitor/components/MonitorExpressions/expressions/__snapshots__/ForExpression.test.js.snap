// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForExpression renders 1`] = `
<div>
  <div
    class="euiText euiText--extraSmall"
  >
    <strong>
      Time range for the last 
    </strong>
    <span
      class="euiToolTipAnchor"
    >
      <div>
        EuiIconMock
      </div>
    </span>
  </div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="max-width:600px"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="width:150px"
    >
      <div
        class="euiFormRow"
        id="bucketValue-form-row-row"
      >
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldNumber"
                min="1"
                name="bucketValue"
                type="number"
                value="1"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
      style="width:150px"
    >
      <div
        class="euiFormControlLayout"
      >
        <div
          class="euiFormControlLayout__childrenWrapper"
        >
          <select
            class="euiSelect"
            id="bucketUnitOfTime"
            name="bucketUnitOfTime"
          >
            <option
              value="m"
            >
              minute(s)
            </option>
            <option
              selected=""
              value="h"
            >
              hour(s)
            </option>
            <option
              value="d"
            >
              day(s)
            </option>
          </select>
          <div
            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
          >
            <span
              class="euiFormControlLayoutCustomIcon"
            >
              <div>
                EuiIconMock
              </div>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
